File: agent/agent.py
#!/usr/bin/env python3
"""
AnyControl Remote — Desktop Agent
=================================

Runs on the computer you want to control from your phone. It captures the
screen, streams JPEG frames to the AnyControl backend over a WebSocket, and
executes the mouse / keyboard events sent back from the mobile app.

Two ways to connect the agent to your account:

  1) Pairing code (recommended, no password on the desktop):
       In the mobile app open "Add Device" to get a 9-digit code + OTP, then:
       python agent.py --server https://YOUR-BACKEND --pair 123456789 --otp ABC123 \
                       --name "My Laptop"

  2) Account login:
       python agent.py --server https://YOUR-BACKEND --email you@mail.com \
                       --password secret --name "My Laptop"

Add --demo to run without a real display (generates synthetic frames and only
logs the control events). Useful for testing the pipeline on a headless server.

Credentials for the device are cached in agent_state.json so subsequent runs
reconnect automatically.
"""

import os
import sys
import json
import time
import base64
import argparse
import asyncio
import logging
from io import BytesIO
from pathlib import Path

import requests
import websockets

logging.basicConfig(level=logging.INFO, format="%(asctime)s [agent] %(message)s")
log = logging.getLogger("agent")

STATE_FILE = Path(__file__).parent / "agent_state.json"
FRAME_INTERVAL = 0.12   # ~8 fps; balances smoothness vs. bandwidth
JPEG_QUALITY = 55


# --------------------------------------------------------------------------- #
# State persistence
# --------------------------------------------------------------------------- #
def load_state() -> dict:
    if STATE_FILE.exists():
        try:
            return json.loads(STATE_FILE.read_text())
        except Exception:
            return {}
    return {}


def save_state(state: dict):
    STATE_FILE.write_text(json.dumps(state, indent=2))


# --------------------------------------------------------------------------- #
# Backend auth / device provisioning
# --------------------------------------------------------------------------- #
def http(server: str, path: str) -> str:
    return server.rstrip("/") + path


def provision(args) -> dict:
    """Return {server, token, device_id} — from cache or by pairing / login."""
    state = load_state()
    if state.get("agent_token") and state.get("device_id") and not args.reset:
        state["server"] = args.server or state.get("server")
        log.info("Using cached device credentials (%s)", state["device_id"])
        return state

    server = args.server or state.get("server")
    if not server:
        log.error("A --server URL is required for first-time setup.")
        sys.exit(1)

    if args.pair:
        r = requests.post(http(server, "/api/devices/pair/claim"), json={
            "code": args.pair, "otp": args.otp or "", "name": args.name,
        }, timeout=20)
        if r.status_code != 200:
            log.error("Pairing failed: %s %s", r.status_code, r.text)
            sys.exit(1)
        data = r.json()
        new_state = {"server": server, "agent_token": data["agent_token"],
                     "device_id": data["device_id"], "name": data["name"]}
        save_state(new_state)
        log.info("Paired successfully as device %s", data["device_id"])
        return new_state

    if args.email and args.password:
        r = requests.post(http(server, "/api/auth/login"),
                          json={"email": args.email, "password": args.password}, timeout=20)
        if r.status_code != 200:
            log.error("Login failed: %s %s", r.status_code, r.text)
            sys.exit(1)
        user_token = r.json()["token"]
        r = requests.post(http(server, "/api/devices/register"),
                          headers={"Authorization": f"Bearer {user_token}"},
                          json={"name": args.name}, timeout=20)
        if r.status_code != 200:
            log.error("Device registration failed: %s %s", r.status_code, r.text)
            sys.exit(1)
        data = r.json()
        new_state = {"server": server, "agent_token": data["agent_token"],
                     "device_id": data["id"], "name": data["name"]}
        save_state(new_state)
        log.info("Registered device %s", data["id"])
        return new_state

    log.error("Provide either --pair/--otp or --email/--password for first setup.")
    sys.exit(1)


# --------------------------------------------------------------------------- #
# Screen capture + input control backends
# --------------------------------------------------------------------------- #
class RealBackend:
    """Real screen capture (mss) and input control (pynput)."""

    def __init__(self):
        import mss
        from pynput.mouse import Controller as Mouse, Button
        from pynput.keyboard import Controller as Keyboard, Key
        self.sct = mss.mss()
        self.monitor = self.sct.monitors[1]
        self.width = self.monitor["width"]
        self.height = self.monitor["height"]
        self.mouse = Mouse()
        self.keyboard = Keyboard()
        self.Button = Button
        self.Key = Key
        from PIL import Image
        self.Image = Image

    def capture(self) -> bytes:
        shot = self.sct.grab(self.monitor)
        img = self.Image.frombytes("RGB", shot.size, shot.bgra, "raw", "BGRX")
        buf = BytesIO()
        img.save(buf, format="JPEG", quality=JPEG_QUALITY)
        return buf.getvalue()

    def _abs(self, nx, ny):
        return (self.monitor["left"] + int(nx * self.width),
                self.monitor["top"] + int(ny * self.height))

    def move(self, nx, ny):
        self.mouse.position = self._abs(nx, ny)

    def click(self, nx, ny, button="left", double=False):
        self.mouse.position = self._abs(nx, ny)
        b = self.Button.right if button == "right" else self.Button.left
        self.mouse.click(b, 2 if double else 1)

    def scroll(self, dy):
        self.mouse.scroll(0, dy)

    def key(self, name):
        special = getattr(self.Key, name, None)
        if special is not None:
            self.keyboard.press(special)
            self.keyboard.release(special)
        elif len(name) == 1:
            self.keyboard.press(name)
            self.keyboard.release(name)

    def type_text(self, text):
        self.keyboard.type(text)

    def hotkey(self, keys):
        pressed = []
        for k in keys:
            key = getattr(self.Key, k, k)
            self.keyboard.press(key)
            pressed.append(key)
        for key in reversed(pressed):
            self.keyboard.release(key)


class DemoBackend:
    """Headless synthetic frames + logged control events (for testing)."""

    def __init__(self):
        from PIL import Image, ImageDraw
        self.Image = Image
        self.ImageDraw = ImageDraw
        self.width = 1280
        self.height = 720
        self.frame = 0

    def capture(self) -> bytes:
        self.frame += 1
        img = self.Image.new("RGB", (self.width, self.height), (15, 15, 15))
        d = self.ImageDraw.Draw(img)
        d.rectangle([0, 0, self.width, 64], fill=(38, 38, 38))
        d.text((24, 22), "AnyControl Remote — DEMO DESKTOP", fill=(255, 109, 0))
        d.text((24, 120), f"Frame #{self.frame}", fill=(242, 242, 242))
        d.text((24, 160), time.strftime("%H:%M:%S"), fill=(163, 163, 163))
        d.ellipse([620, 340, 660, 380], outline=(255, 109, 0), width=3)
        buf = BytesIO()
        img.save(buf, format="JPEG", quality=JPEG_QUALITY)
        return buf.getvalue()

    def move(self, nx, ny):
        log.info("move -> (%.3f, %.3f)", nx, ny)

    def click(self, nx, ny, button="left", double=False):
        log.info("click %s%s -> (%.3f, %.3f)", button, " x2" if double else "", nx, ny)

    def scroll(self, dy):
        log.info("scroll %s", dy)

    def key(self, name):
        log.info("key %s", name)

    def type_text(self, text):
        log.info("type %r", text)

    def hotkey(self, keys):
        log.info("hotkey %s", "+".join(keys))


# --------------------------------------------------------------------------- #
# Main loop
# --------------------------------------------------------------------------- #
async def run(state: dict, backend, demo: bool):
    ws_base = state["server"].replace("http://", "ws://").replace("https://", "wss://")
    url = f"{ws_base.rstrip('/')}/api/ws/agent?token={state['agent_token']}"

    while True:
        try:
            async with websockets.connect(url, max_size=None, ping_interval=20) as ws:
                log.info("Connected to relay. Waiting for viewer...")
                viewer_active = {"on": False}

                async def stream():
                    while True:
                        if viewer_active["on"]:
                            data = backend.capture()
                            b64 = base64.b64encode(data).decode("ascii")
                            await ws.send(json.dumps({
                                "type": "frame", "data": b64,
                                "w": backend.width, "h": backend.height,
                            }))
                            await asyncio.sleep(FRAME_INTERVAL)
                        else:
                            await asyncio.sleep(0.25)

                async def recv():
                    async for raw in ws:
                        try:
                            msg = json.loads(raw)
                        except Exception:
                            continue
                        handle_control(backend, msg, viewer_active)

                await asyncio.gather(stream(), recv())
        except Exception as e:
            log.warning("Connection lost (%s). Reconnecting in 3s...", e)
            await asyncio.sleep(3)


def handle_control(backend, msg: dict, viewer_active: dict):
    t = msg.get("type")
    if t == "viewer":
        viewer_active["on"] = bool(msg.get("active"))
        log.info("Viewer %s", "attached" if viewer_active["on"] else "detached")
    elif t == "move":
        backend.move(msg["x"], msg["y"])
    elif t == "click":
        backend.click(msg["x"], msg["y"], msg.get("button", "left"), msg.get("double", False))
    elif t == "scroll":
        backend.scroll(msg.get("dy", 0))
    elif t == "key":
        backend.key(msg["key"])
    elif t == "text":
        backend.type_text(msg["text"])
    elif t == "hotkey":
        backend.hotkey(msg.get("keys", []))


def main():
    ap = argparse.ArgumentParser(description="AnyControl Remote desktop agent")
    ap.add_argument("--server", default=os.environ.get("ANYCONTROL_SERVER"))
    ap.add_argument("--name", default=os.environ.get("ANYCONTROL_NAME", "My Computer"))
    ap.add_argument("--pair", help="9-digit pairing code from the mobile app")
    ap.add_argument("--otp", help="OTP shown with the pairing code")
    ap.add_argument("--email")
    ap.add_argument("--password")
    ap.add_argument("--demo", action="store_true", help="headless synthetic frames")
    ap.add_argument("--reset", action="store_true", help="ignore cached credentials")
    args = ap.parse_args()

    state = provision(args)
    backend = DemoBackend() if args.demo else RealBackend()
    log.info("Screen %dx%d | mode=%s", backend.width, backend.height,
             "demo" if args.demo else "live")
    try:
        asyncio.run(run(state, backend, args.demo))
    except KeyboardInterrupt:
        log.info("Agent stopped.")


if __name__ == "__main__":
    main()

File: agent/src/agent.ts
import fs from 'fs';
import path from 'path';
import axios from 'axios';
import WebSocket from 'ws';
import { Command } from 'commander';
import dotenv from 'dotenv';
import jpeg from 'jpeg-js';

// Load environment variables
const ROOT_DIR = path.join(__dirname, '..');
dotenv.config({ path: path.join(ROOT_DIR, '.env') });

const STATE_FILE = path.join(ROOT_DIR, 'agent_state.json');
const FRAME_INTERVAL = 120; // ms (~8 fps)

interface State {
  server?: string;
  agent_token?: string;
  device_id?: string;
  name?: string;
}

function loadState(): State {
  if (fs.existsSync(STATE_FILE)) {
    try {
      return JSON.parse(fs.readFileSync(STATE_FILE, 'utf-8'));
    } catch (err) {
      return {};
    }
  }
  return {};
}

function saveState(state: State) {
  fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2), 'utf-8');
}

// REST helper
function http(server: string, endpoint: string): string {
  return server.replace(/\/$/, '') + endpoint;
}

// Provisioning logic
async function provision(args: any): Promise<State> {
  const state = loadState();
  if (state.agent_token && state.device_id && !args.reset) {
    state.server = args.server || state.server;
    console.log(`Using cached device credentials (${state.device_id})`);
    return state;
  }

  const server = args.server || state.server;
  if (!server) {
    console.error('A --server URL is required for first-time setup.');
    process.exit(1);
  }

  if (args.pair) {
    try {
      console.log(`Attempting to claim pairing code: ${args.pair}...`);
      const response = await axios.post(http(server, '/api/devices/pair/claim'), {
        code: args.pair,
        otp: args.otp || '',
        name: args.name,
      });

      const data = response.data;
      const newState: State = {
        server,
        agent_token: data.agent_token,
        device_id: data.device_id,
        name: data.name,
      };
      saveState(newState);
      console.log(`Paired successfully as device ${data.device_id}`);
      return newState;
    } catch (err: any) {
      console.error(`Pairing failed: ${err.response?.status} ${JSON.stringify(err.response?.data)}`);
      process.exit(1);
    }
  }

  if (args.email && args.password) {
    try {
      console.log(`Attempting account login for ${args.email}...`);
      const loginRes = await axios.post(http(server, '/api/auth/login'), {
        email: args.email,
        password: args.password,
      });
      const userToken = loginRes.data.token;

      console.log('Registering device...');
      const deviceRes = await axios.post(
        http(server, '/api/devices/register'),
        { name: args.name },
        { headers: { Authorization: `Bearer ${userToken}` } }
      );

      const data = deviceRes.data;
      const newState: State = {
        server,
        agent_token: data.agent_token,
        device_id: data.id,
        name: data.name,
      };
      saveState(newState);
      console.log(`Registered device successfully as ${data.id}`);
      return newState;
    } catch (err: any) {
      console.error(`Registration failed: ${err.response?.status} ${JSON.stringify(err.response?.data)}`);
      process.exit(1);
    }
  }

  }

  // If no provisioning parameters are provided, default to QR code pairing
  console.log("No credentials provided. Initializing terminal QR code pairing...");
  try {
    const qrcode = require('qrcode-terminal');
    console.log(`Contacting server ${server} for pairing code...`);
    const initRes = await axios.post(http(server, '/api/devices/pair/agent-new'), {
      name: args.name
    });
    const { code, otp } = initRes.data;

    // We encode the pairing details into a JSON payload
    const qrPayload = JSON.stringify({ code, otp });
    
    console.log('\n======================================================');
    console.log('To link this laptop, scan the QR code below on your phone:');
    console.log('======================================================\n');
    
    qrcode.generate(qrPayload, { small: true });

    console.log(`Pairing Code: ${code}  |  OTP: ${otp}`);
    console.log('Waiting for you to scan this QR code inside the web app...\n');

    // Poll until claimed
    while (true) {
      try {
        const checkRes = await axios.post(http(server, '/api/devices/pair/agent-check'), { code, otp });
        if (checkRes.data.status === 'claimed') {
          const data = checkRes.data;
          const newState: State = {
            server,
            agent_token: data.agent_token,
            device_id: data.device_id,
            name: data.name,
          };
          saveState(newState);
          console.log(`\nSuccess! Registered device successfully as ${data.device_id}\n`);
          return newState;
        }
      } catch (checkErr) {
        console.error('\nPairing failed or session expired. Please restart the agent.');
        process.exit(1);
      }
      await new Promise(r => setTimeout(r, 2000));
    }
  } catch (err: any) {
    console.error(`QR Code pairing setup failed: ${err.message || err}`);
    process.exit(1);
  }
}

// --------------------------------------------------------------------------- //
// Capturing and input simulation backends
// --------------------------------------------------------------------------- //

class DemoBackend {
  width = 1280;
  height = 720;
  frame = 0;

  capture(): Buffer {
    this.frame++;
    const frameData = Buffer.alloc(this.width * this.height * 4);

    // Draw background (dark gray: 15, 15, 15)
    for (let i = 0; i < frameData.length; i += 4) {
      frameData[i] = 15;     // R
      frameData[i + 1] = 15; // G
      frameData[i + 2] = 15; // B
      frameData[i + 3] = 255; // A
    }

    // Draw simulated active moving block to verify relay framerate
    const rectSize = 100;
    const rectX = Math.floor((this.width - rectSize) / 2 + Math.cos(this.frame * 0.1) * 300);
    const rectY = Math.floor((this.height - rectSize) / 2 + Math.sin(this.frame * 0.1) * 200);

    for (let y = rectY; y < rectY + rectSize; y++) {
      if (y < 0 || y >= this.height) continue;
      for (let x = rectX; x < rectX + rectSize; x++) {
        if (x < 0 || x >= this.width) continue;
        const idx = (y * this.width + x) * 4;
        frameData[idx] = 255;     // Orange R
        frameData[idx + 1] = 109; // Orange G
        frameData[idx + 2] = 0;   // Orange B
      }
    }

    const rawImageData = {
      data: frameData,
      width: this.width,
      height: this.height,
    };
    const jpegImageData = jpeg.encode(rawImageData, 55);
    return jpegImageData.data;
  }

  move(nx: number, ny: number) {
    console.log(`[demo] move -> (${nx.toFixed(3)}, ${ny.toFixed(3)})`);
  }

  click(nx: number, ny: number, button = 'left', double = false) {
    console.log(`[demo] click ${button}${double ? ' x2' : ''} -> (${nx.toFixed(3)}, ${ny.toFixed(3)})`);
  }

  scroll(dy: number) {
    console.log(`[demo] scroll ${dy}`);
  }

  key(name: string) {
    console.log(`[demo] key ${name}`);
  }

  typeText(text: string) {
    console.log(`[demo] type ${JSON.stringify(text)}`);
  }

  hotkey(keys: string[]) {
    console.log(`[demo] hotkey ${keys.join('+')}`);
  }
}

class RealBackend {
  width = 1920;
  height = 1080;
  private nutMouse: any;
  private nutKeyboard: any;
  private nutKey: any;
  private nutButton: any;
  private screenshot: any;

  constructor() {
    try {
      this.nutMouse = require('@nut-tree-fork/nut-js').mouse;
      this.nutKeyboard = require('@nut-tree-fork/nut-js').keyboard;
      this.nutKey = require('@nut-tree-fork/nut-js').Key;
      this.nutButton = require('@nut-tree-fork/nut-js').Button;
      this.screenshot = require('screenshot-desktop');
    } catch (err) {
      console.error('Failed to load real input automation dependencies. Forcing fallback to DemoBackend.', err);
      throw err;
    }
  }

  async capture(): Promise<Buffer> {
    return await this.screenshot({ format: 'jpeg' });
  }

  async move(nx: number, ny: number) {
    const targetX = Math.floor(nx * this.width);
    const targetY = Math.floor(ny * this.height);
    const { Point } = require('@nut-tree-fork/nut-js');
    await this.nutMouse.setPosition(new Point(targetX, targetY));
  }

  async click(nx: number, ny: number, button = 'left', double = false) {
    await this.move(nx, ny);
    const btn = button === 'right' ? this.nutButton.RIGHT : this.nutButton.LEFT;
    if (double) {
      await this.nutMouse.doubleClick(btn);
    } else {
      await this.nutMouse.click(btn);
    }
  }

  async scroll(dy: number) {
    if (dy > 0) {
      await this.nutMouse.scrollUp(Math.abs(dy) * 100);
    } else if (dy < 0) {
      await this.nutMouse.scrollDown(Math.abs(dy) * 100);
    }
  }

  async key(name: string) {
    const keyEnum = (this.nutKey as any)[name.toUpperCase()] || name;
    await this.nutKeyboard.pressKey(keyEnum);
    await this.nutKeyboard.releaseKey(keyEnum);
  }

  async typeText(text: string) {
    await this.nutKeyboard.type(text);
  }

  async hotkey(keys: string[]) {
    const mappedKeys = keys.map(k => (this.nutKey as any)[k.toUpperCase()] || k);
    await this.nutKeyboard.pressKey(...mappedKeys);
    await this.nutKeyboard.releaseKey(...mappedKeys);
  }
}

// --------------------------------------------------------------------------- //
// Websocket streaming execution loop
// --------------------------------------------------------------------------- //

async function run(state: State, backend: any, demo: boolean) {
  const wsBase = state.server!
    .replace('http://', 'ws://')
    .replace('https://', 'wss://')
    .replace(/\/$/, '');
  const url = `${wsBase}/api/ws/agent?token=${state.agent_token}`;

  while (true) {
    try {
      console.log(`Connecting to WebSocket relay: ${url}`);
      await new Promise<void>((resolve, reject) => {
        const ws = new WebSocket(url);
        let viewerActive = { on: false };
        let streamTimer: NodeJS.Timeout | null = null;

        ws.on('open', () => {
          console.log('Connected to relay. Waiting for viewer...');

          // Run streaming loop
          streamTimer = setInterval(async () => {
            if (viewerActive.on) {
              try {
                const imgData = await backend.capture();
                const b64 = imgData.toString('base64');
                ws.send(JSON.stringify({
                  type: 'frame',
                  data: b64,
                  w: backend.width,
                  h: backend.height,
                }));
              } catch (err) {
                console.error('Capture/send frame failed:', err);
              }
            }
          }, FRAME_INTERVAL);
        });

        ws.on('message', async (data) => {
          try {
            const msg = JSON.parse(data.toString());
            await handleControl(backend, msg, viewerActive);
          } catch (err) {
            // ignore malformed packets
          }
        });

        ws.on('close', (code, reason) => {
          console.warn(`WebSocket disconnected: ${code} - ${reason.toString()}`);
          if (streamTimer) clearInterval(streamTimer);
          reject(new Error('Connection closed'));
        });

        ws.on('error', (err) => {
          console.error('WebSocket connection error:', err);
          if (streamTimer) clearInterval(streamTimer);
          reject(err);
        });
      });
    } catch (err) {
      console.log('Reconnecting in 3 seconds...');
      await new Promise(r => setTimeout(r, 3000));
    }
  }
}

async function handleControl(backend: any, msg: any, viewerActive: any) {
  const t = msg.type;
  if (t === 'viewer') {
    viewerActive.on = !!msg.active;
    console.log(`Viewer ${viewerActive.on ? 'attached' : 'detached'}`);
  } else if (t === 'move') {
    await backend.move(msg.x, msg.y);
  } else if (t === 'click') {
    await backend.click(msg.x, msg.y, msg.button || 'left', !!msg.double);
  } else if (t === 'scroll') {
    await backend.scroll(msg.dy || 0);
  } else if (t === 'key') {
    await backend.key(msg.key);
  } else if (t === 'text') {
    await backend.typeText(msg.text);
  } else if (t === 'hotkey') {
    await backend.hotkey(msg.keys || []);
  }
}

// --------------------------------------------------------------------------- //
// CLI Entrypoint
// --------------------------------------------------------------------------- //

async function main() {
  const program = new Command();
  program
    .description('AnyControl Remote Desktop Agent in Node.js')
    .option('--server <url>', 'AnyControl server URL', process.env.ANYCONTROL_SERVER)
    .option('--name <name>', 'Device registration name', process.env.ANYCONTROL_NAME || 'My Desktop')
    .option('--pair <code>', '9-digit pairing code from mobile app')
    .option('--otp <otp>', 'OTP paired code validation')
    .option('--email <email>', 'Account email')
    .option('--password <password>', 'Account password')
    .option('--demo', 'Enable headless synthetic frame generation', false)
    .option('--reset', 'Ignore cached authentication state', false);

  program.parse(process.argv);
  const options = program.opts();

  const state = await provision(options);

  let backend: any;
  if (options.demo) {
    console.log('Starting agent in DEMO mode');
    backend = new DemoBackend();
  } else {
    try {
      console.log('Starting agent in LIVE mode');
      backend = new RealBackend();
    } catch (err) {
      console.warn('Fallback to DEMO mode due to environment missing automation libs');
      backend = new DemoBackend();
      options.demo = true;
    }
  }

  console.log(`Screen resolution: ${backend.width}x${backend.height} | Mode: ${options.demo ? 'demo' : 'live'}`);

  process.on('SIGINT', () => {
    console.log('Agent stopped.');
    process.exit(0);
  });

  await run(state, backend, options.demo);
}

main().catch(err => {
  console.error('Unhandled fatal exception:', err);
  process.exit(1);
});
